tensorflow/core/kernels/resource_variable_ops.cc

```diff
@@ -900,8 +900,33 @@ REGISTER_KERNEL_BUILDER(Name("ResourceGather")
 
 #endif  // GOOGLE_CUDA || TENSORFLOW_USE_ROCM
 
+#if TENSORFLOW_USE_DIRECTML
+#define REGISTER_GATHER_DML(type) REGISTER_GATHER_ALL_INDICES(CPU, type)
+
+TF_CALL_DML_FLOAT_TYPES(REGISTER_GATHER_DML);
+
+// Variant objects themselves sit on CPU, even if they contain data
+// pointing to a device.
+REGISTER_KERNEL_BUILDER(Name("ResourceGather")
+                            .Device(DEVICE_DML)
+                            .HostMemory("resource")
+                            .HostMemory("indices")
+                            .TypeConstraint<Variant>("dtype")
+                            .TypeConstraint<int32>("Tindices"),
+                        ResourceGatherOp<CPUDevice, Variant, int32>)
+REGISTER_KERNEL_BUILDER(Name("ResourceGather")
+                            .Device(DEVICE_DML)
+                            .HostMemory("resource")
+                            .HostMemory("indices")
+                            .TypeConstraint<Variant>("dtype")
+                            .TypeConstraint<int64>("Tindices"),
+                        ResourceGatherOp<CPUDevice, Variant, int64>)
```

Comment:
1. This won't work. Resource kernels require their own DML implementation, we can't just register them as `CPUDevice` and call it a day. Notice how the GPU kernel registration uses `GPUDevice` and not `CPUDevice`. You should look at resource operators that we alteady implemented in DML (e.g. ResourceApplyAdam).
2. If you still want to implement the resource operators, you should put them in their own PR. Resource operators and tensorlist operators have nothing in common, so this PR is now slightly misleading.